nexus/src/db/subnet_allocation.rs

```diff
@@ -41,12 +42,173 @@ fn generate_last_address_offset(subnet: &ipnetwork::IpNetwork) -> i64 {
+                )
+            },
+            SubnetError::External(e) => e,
+        }
+    }
+}
+
+/// Generate a subquery that selects any overlapping address ranges of the same
+/// type as the input IP subnet.
+///
+/// This generates a query that, in full, looks like:
+///
+/// ```sql
+/// SELECT
+///     <ip>
+/// FROM
+///     vpc_subnet
+/// WHERE
+///     vpc_id = <vpc_id> AND
+///     time_deleted IS NULL AND
+///     inet_contains_or_equals(ipv*_block, <ip>)
+/// LIMIT 1
+/// ```
+///
+/// The input may be either an IPv4 or IPv6 subnet, and the corresponding column
+/// is compared against. Note that the exact input IP range is returned on
+/// purpose.
+fn push_select_overlapping_ip_range(
+    mut out: AstPass<Pg>,
+    vpc_id: &Uuid,
+    ip: &ipnetwork::IpNetwork,
+) -> diesel::QueryResult<()> {
+    use crate::db::schema::vpc_subnet::dsl;
+    out.push_sql("SELECT ");
+    out.push_bind_param::<sql_types::Inet, ipnetwork::IpNetwork>(ip)?;
+    out.push_sql(" FROM ");
+    dsl::vpc_subnet.from_clause().walk_ast(out.reborrow())?;
+    out.push_sql(" WHERE ");
+    out.push_identifier(dsl::vpc_id::NAME)?;
+    out.push_sql(" = ");
+    out.push_bind_param::<sql_types::Uuid, Uuid>(vpc_id)?;
+    out.push_sql(" AND ");
+    out.push_identifier(dsl::time_deleted::NAME)?;
+    out.push_sql(" IS NULL AND inet_contains_or_equals(");
+    if ip.is_ipv4() {
+        out.push_identifier(dsl::ipv4_block::NAME)?;
+    } else {
+        out.push_identifier(dsl::ipv6_block::NAME)?;
+    }
+    out.push_sql(", ");
+    out.push_bind_param::<sql_types::Inet, ipnetwork::IpNetwork>(ip)?;
+    out.push_sql(")");
+    Ok(())
+}
+
+/// Generate a subquery that returns NULL if there is an overlapping IP address
+/// range of any type.
+///
+/// This specifically generates a query that looks like:
+///
+/// ```sql
+/// SELECT NULLIF(
+///     <ip>,
+///     select_overlapping_ip_range(<vpc_id>, <ip>)
+/// )
+/// ```
+///
+/// The `NULLIF` function returns NULL if those two expressions are equal, and
+/// the first expression otherwise. That is, this returns NULL if there exists
+/// an overlapping IP range already in the VPC Subnet table, and the requested
+/// IP range if not.
+fn push_null_if_overlapping_ip_range(
+    mut out: AstPass<Pg>,
+    vpc_id: &Uuid,
+    ip: &ipnetwork::IpNetwork,
+) -> diesel::QueryResult<()> {
+    out.push_sql("SELECT NULLIF(");
+    out.push_bind_param::<sql_types::Inet, ipnetwork::IpNetwork>(ip)?;
+    out.push_sql(", (");
+    push_select_overlapping_ip_range(out.reborrow(), vpc_id, ip)?;
+    out.push_sql("))");
+    Ok(())
+}
+
+// Use the not-NULL constraint on these columns to detect and distinguish
+/// WITH
+/// candidate(...) AS <provided values>,
+/// candidate_ipv4(ipv4_block) AS push_null_if_overlapping_ip_range,
+/// candidate_ipv6(ipv6_block) AS push_null_if_overlapping_ip_range
+/// SELECT * FROM candidate, candidate_ipv4, candidate_ivp6;
```

Comment:
Is this comment supposed to be attached somewhere?

Reply:
Stray comment from a rebase, will remove.